applications/nrf_desktop/src/modules/usb_state.c

```diff
@@ -126,6 +140,151 @@ enum {
+		}
+	}
+
+	/* Should not happen. */
+	__ASSERT_NO_MSG(report_id != REPORT_ID_COUNT);
+
+	return report_id;
+}
+
+static struct usb_hid_buf *usb_hid_buf_alloc(struct usb_hid_device *usb_hid, const uint8_t *data,
+					     size_t size)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(usb_hid->report_bufs); i++) {
+		struct usb_hid_buf *r = &usb_hid->report_bufs[i];
+
+		if (!(r->status_bm & USB_HID_BUF_ALLOCATED)) {
+			__ASSERT_NO_MSG(!(r->status_bm & USB_HID_BUF_SENDING));
+			__ASSERT_NO_MSG(!(r->status_bm & USB_HID_BUF_BOOT_REPORT_FORMAT));
+			__ASSERT_NO_MSG(sizeof(r->data) >= size);
+
+			memcpy(r->data, data, size);
+			r->size = size;
+			r->status_bm |= USB_HID_BUF_ALLOCATED;
+			return r;
+		}
+	}
+
+	return NULL;
+}
+
+static void usb_hid_buf_free(struct usb_hid_buf *report_buf)
+{
+	report_buf->status_bm = 0;
+}
+
+static struct usb_hid_buf *usb_hid_buf_find(struct usb_hid_device *usb_hid, uint8_t status_bm)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(usb_hid->report_bufs); i++) {
+		struct usb_hid_buf *r = &usb_hid->report_bufs[i];
+
+		if ((status_bm & r->status_bm) == status_bm) {
+			return r;
+		}
+	}
+
+	return NULL;
+}
+
+static bool is_hid_boot_report(uint8_t report_id)
+{
+	return ((report_id == REPORT_ID_BOOT_MOUSE) || (report_id == REPORT_ID_BOOT_KEYBOARD));
+}
+
+static bool is_hid_boot_report_supported(uint8_t report_id)
+{
+	__ASSERT_NO_MSG(is_hid_boot_report(report_id));
+
+	return ((IS_ENABLED(CONFIG_DESKTOP_HID_BOOT_INTERFACE_MOUSE) &&
+		 (report_id == REPORT_ID_BOOT_MOUSE)) ||
+		(IS_ENABLED(CONFIG_DESKTOP_HID_BOOT_INTERFACE_KEYBOARD) &&
+		 (report_id == REPORT_ID_BOOT_KEYBOARD)));
+}
+
+static bool can_send_hid_report(struct usb_hid_device *usb_hid, uint8_t report_id)
+{
+	if (!usb_hid->enabled) {
+		LOG_WRN("Cannot send report: USB not connected");
+		return false;
+	}
+
+	if (usb_hid->hid_protocol == HID_PROTOCOL_BOOT) {
+		if (!is_hid_boot_report(report_id)) {
+			LOG_WRN("Cannot send report: incompatible boot/report mode");
+			return false;
+		}
+
+		if (!is_hid_boot_report_supported(report_id)) {
+			LOG_WRN("Cannot send report: unsupported boot report");
+			return false;
+		}
+	} else {
+		__ASSERT_NO_MSG(usb_hid->hid_protocol == HID_PROTOCOL_REPORT);
+
+		if (is_hid_boot_report(report_id)) {
+			LOG_WRN("Cannot send report: incompatible boot/report mode");
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static void usb_hid_buf_send(struct usb_hid_device *usb_hid, struct usb_hid_buf *buf)
+{
+	uint8_t report_id = usb_hid_buf_get_report_id(buf);
+
+	buf->status_bm |= USB_HID_BUF_SENDING;
+
+	if (!can_send_hid_report(usb_hid, report_id)) {
+		report_sent(usb_hid, buf, true);
+		return;
+	}
+
+	uint8_t *data = buf->data;
+	size_t size = buf->size;
+
+	if (is_hid_boot_report(report_id)) {
+		/* Omit report ID for HID boot reports. Keep proper memory alignment. */
+		for (size_t i = 1; i < size; i++) {
+			data[i - 1] = data[i];
+		}
+		size--;
+		buf->status_bm |= USB_HID_BUF_BOOT_REPORT_FORMAT;
```

Comment:
you seem to use the usb_hid_buf_get_report_id function, which depends on this flag setting

Reply:
This one is a bit tricky. Since a HID boot report needs to be passed to USB stack without being prepended with report ID and data buffer alignment is still required, we need another way to store information that a `usb_hid_buf` contains a HID boot report. The `USB_HID_BUF_BOOT_REPORT_FORMAT` flag means that data buffer was "reformatted" to follow the boot protocol format (basically report ID was removed and content was shifted by one). `usb_hid_buf_get_report_id` needs to use special handling to get the report ID from this particular buffer. The flag is set together with "reformatting" operation to ensure consistency.

Reply:
I know but you seem to call the `usb_hid_buf_get_report_id` this function before you set this boot formatter

Reply:
Then we still use the "regular" report format (since report is not yet "reformatted"). Would work fine :)

Reply:
thanks, discussed offline